core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -117,9 +126,23 @@
   }
 
   private final class ContextCancellationListener implements CancellationListener {
+    private Listener<RespT> observer;
+
+    private void setObserver(Listener<RespT> observer) {
+      this.observer = observer;
+    }
+
     @Override
-    public void cancelled(Context context) {
-      stream.cancel(statusFromCancelled(context));
+    public void cancelled(final Context context) {
+      if (context.getDeadline() == null || !context.getDeadline().isExpired()) {
+        stream.cancel(statusFromCancelled(context));
+      } else {
+        if (observer != null) {
+          Status status = statusFromCancelled(context);
+          deadlineCancellationSendToServerFuture = startDeadlineSendCancelToServerTimer(status);
+          executeCloseObserverInContext(observer, status);
```

Comment:
These two lines are repeated in `DeadlineExceededNotifyApplicationTimer`. Since this is the core logic of delayed cancellation, it should be in its own method (`delayedCancelOnDeadlineExceeded(status, observer)`), and the two cases just call it.